### DISCUSSION TITLE
Directly call cowboy REST handler inside app

### DISCUSSION THREAD

**Discussion by bever1337** [ninenines/cowboy#1642] [Q&A] [answered]
Howdy! Is there a way to directly execute a REST handler? For example,a REST service uses hypermedia with linking, ex [JSON HAL](https://datatracker.ietf.org/doc/html/draft-kelly-json-hal-08#section-8.3). When a response links to another resource, the server optionally embeds that second resource into the first response.

I would like to use a middleware approach, but I'm not sure how to wait for the resolution of additional requests without using HTTP (or whatever protocol that cowboy+handler is configured to use).

Alternatively, I know handler `foo` can call exports from handler `bar`, but this defeats the control flow that cowboy provides. I could also create an additional service that is responsible for providing resources and generating pages, but it feels like duplication of effort.

Thank y'all for the consideration

**Comment by essen** [comment, marked as answer]
There currently isn't. Part of the reason is that REST handlers are still very closely tied to HTTP, whereas to do what you want it would be more appropriate to have protocol-generic resource. This is what I've been experimenting on in https://github.com/ninenines/farwest but it is incomplete. Some parts of this experimentation will make it into Cowboy 3.0 (new router, at the very least, so that we can find the resources).

As it currently stands, you probably should run the handler in a separate process, since `cowboy_rest` will send the response as a message. So what you need to use is `cowboy_stream` and let Cowboy handle it from there. Then your own code needs to respond to messages and provide them back to `cowboy_stream`, which will return commands including the response you seek.

**Comment by bever1337** [comment]
Thank you very much for the prompt reply. This is a huge help.

**Reply by essen** [reply]
Happy to see what you end up with as it'll help me for further experiments or an actual implementation.
